src/charmcraft/charmcraft.js

```diff
@@ -0,0 +1,109 @@
+const core = require("@actions/core");
+const exec = require("@actions/exec");
+const fs = require("fs");
+const yaml = require("js-yaml");
+const glob = require("@actions/glob");
+
+/* eslint-disable camelcase */
+
+class Charmcraft {
+  constructor() {
+    this.uploadImage = core.getInput("upload-image").toLowerCase() === "true";
+    core.exportVariable("CHARMCRAFT_AUTH", core.getInput("credentials"));
+  }
+
+  async uploadResources() {
+    let resourceInfo = "resources:\n";
+
+    const { name, images } = this.metadata();
+    const flags = await Promise.all(
+      images.map(async ([resource_name, resource_image]) => {
+        await this.uploadResource(resource_image, name, resource_name);
+        const { flag, info } = await this.buildResourceFlag(
+          name,
+          resource_name,
+          resource_image
+        );
+        resourceInfo += info;
+        return flag;
+      })
+    );
+    return { flags, resourceInfo };
+  }
+
+  async uploadResource(resource_image, name, resource_name) {
+    if (!this.uploadImage) {
+      core.warning(
+        "No resources where uploaded as part of this build.\n" +
+          `If you wish to upload the OCI image, set 'upload-image' to 'true'`
+      );
+      return;
+    }
+
+    await exec.exec("docker", ["pull", resource_image]);
+    await exec.exec("charmcraft", [
+      "upload-resource",
+      "--quiet",
+      name,
+      resource_name,
+      "--image",
+      resource_image,
+    ]);
+  }
+
+  async buildResourceFlag(name, resource_name, resource_image) {
+    const result = await exec.getExecOutput("charmcraft", [
+      "resource-revisions",
+      name,
+      resource_name,
+    ]);
+
+    const revision = result.stdout.split("\n")[1].split(" ")[0];
+
+    return {
+      flag: `--resource=${resource_name}:${revision}`,
+      info:
+        `    -  ${resource_name}: ${resource_image}\n` +
+        `       resource-revision: ${revision}\n`,
+    };
+  }
+
+  metadata() {
+    const metadata = yaml.load(fs.readFileSync("metadata.yaml"));
+    const charmName = metadata.name;
+
+    const images = Object.entries(metadata.resources || {})
+      .filter(([, res]) => res.type === "oci-image")
+      .map(([name, res]) => [name, res["upstream-source"]]);
+    return { images, name: charmName };
+  }
+
+  async pack() {
+    await exec.exec("charmcraft", ["pack", "--destructive-mode", "--quiet"]);
+  }
+
+  async upload(channel, revisions) {
+    const globber = await glob.create("./*.charm");
+    const paths = await globber.glob();
+
+    const charmRevisions = await Promise.all(
+      paths.map(async (path) => this._upload_charm(channel, path, revisions))
+    );
+    return charmRevisions;
+  }
+
+  async _upload_charm(channel, path, revisions) {
+    const result = await exec.getExecOutput("charmcraft", [
+      "upload",
+      "--quiet",
+      "--release",
```

Comment:
This could be optional? Not sure we will always want to release? Kinda subjective though so not a blocker

Reply:
I suggest keeping it like this until someone expresses a use case where it would make sense to opt-out. 

Mainly because it would be an excellent prompt for me to pick their brain about their rationale. Changing it to an opt-out itself is a trivial change.